api/client/src/test/java/org/projectnessie/client/http/BaseTestHttpClient.java

```diff
@@ -479,31 +480,107 @@ static Stream<Status> testHttpResponses() {
   }
 
   @ParameterizedTest
-  @MethodSource
+  @MethodSource("httpResponses")
   void testHttpResponses(Status status) throws Exception {
-    // HTTP/304 defines that no response body must be sent
-    assumeThat(status).isNotEqualTo(Status.NOT_MODIFIED);
+    ExampleBean expected = new ExampleBean("x", 1, NOW);
+    HttpTestServer.RequestHandler handler =
+        (req, resp) -> {
+          resp.setStatus(status.getCode());
+          if (status != Status.NO_CONTENT && status != Status.NOT_MODIFIED) {
+            resp.setContentType("application/json");
+            try (OutputStream os = resp.getOutputStream()) {
+              MAPPER.writeValue(os, expected);
+            }
+          }
+        };
+    try (HttpTestServer server = new HttpTestServer("/a/b", handler)) {
```

Comment:
Hm - the overall test duration increased quite a bit (I've actually made it worse than before with #9062).

I think, we can create the test server once per test class and reuse it by having a `static AtomicReference<HttpTestServer.RequestHandler>` and then re-delegate:
```java
    static final AtomicReference<HttpTestServer.RequestHandler> requestHandler = new AtomicReference<>();
    static HttpTestServer server;
    @BeforeAll
    static void setHttpServer() {
      server = new HttpTestServer("/a/b", (req, resp) -> {
        handler.get().handle(req, resp);
      });
    }
    @AfterAll
    static void shutdown() {
      server.stop();
    }
```

That should make the tests run faster.  (We run those for all clients and a bunch of Jackson versions.)

Reply:
I improved things a bit but there is something weird with the Apache client: it's much slower than the other two. Not sure if this is a regression introduced in this PR though.

Reply:
On main:
<img width="364" alt="Screenshot 2024-07-12 at 12 27 02" src="https://github.com/user-attachments/assets/7412e791-ad7b-4086-b0b7-bce7d2d25262">

On this branch, removing the new tests for an apples to apples comparison, but including the changes to share the same client and server between tests:

<img width="354" alt="Screenshot 2024-07-12 at 12 34 04" src="https://github.com/user-attachments/assets/df8d0baf-999c-4649-82ec-8bee3c50221d">

It's roughly the same times, but indeed the Apache client is performing worse than the others. 

It seems to "hang" when processing some HTTP codes like 429 or 503. Looking a bit in the javadocs, I found the [`DefaultBackoffStrategy`](https://hc.apache.org/httpcomponents-client-5.1.x/current/httpclient5/apidocs/org/apache/hc/client5/http/impl/classic/DefaultBackoffStrategy.html) class:

> This [ConnectionBackoffStrategy](https://hc.apache.org/httpcomponents-client-5.1.x/current/httpclient5/apidocs/org/apache/hc/client5/http/classic/ConnectionBackoffStrategy.html) backs off either for a raw network socket or connection timeout or if the server explicitly sends a 429 (Too Many Requests) or a 503 (Service Unavailable) response.

So, this might explain the differences.

Reply:
Oh! Backoff :+1: